packages/stylist-macros/src/inline/output/maybe_static.rs

```diff
@@ -0,0 +1,153 @@
+//! This module implements a type abstractly tracking in what kind of expression context
+//! an item appears. This information is leverage to provide improved performance and
+//! static caching of parts of the generated output.
+use std::{
+    iter::FromIterator,
+    ops::{BitAnd, BitAndAssign},
+};
+
+use proc_macro2::TokenStream;
+use quote::quote;
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub enum ExpressionContext {
+    // ```
+    // let width = 500;
+    // style! { width: ${width}; }
+    // //               ^^^^^^ dynamic expression, can't wrap style in Lazy
+    // ```
+    Dynamic,
+    // ```
+    // style! { width: 500px; }
+    // //       ------------- everything is static, do wrap style in Lazy
+    // ```
+    Static,
+    // TODO: we can probably avoid a few allocations if we track which parts
+    // of the ast can be constructed statically (with const methods), which is
+    // even stronger than constructing it in the global context in a Lazy.
+    // Should you decide to implement this, keep in mind to change Self::MAX
+    // and adjust the generation of cow-vec tokens. Also check the usages of
+    // MaybeStatic::statick if they can be upgraded to Const.
+    // Const,
+}
+
+impl ExpressionContext {
+    const MAX: Self = Self::Static;
+}
+
+impl BitAnd for ExpressionContext {
+    type Output = Self;
+
+    fn bitand(self, rhs: Self) -> Self::Output {
+        self.min(rhs)
+    }
+}
+
+impl BitAndAssign for ExpressionContext {
+    fn bitand_assign(&mut self, rhs: Self) {
+        *self = *self & rhs
+    }
+}
+
+// Used e.g. to decide whether a sheet can be statically cached in a Lazy or must be
+// created everytime anew.
+pub struct MaybeStatic<T> {
+    value: T,
+    context: ExpressionContext,
```

Comment:
I agree that using the `&` might not be ideal, I changed it after previously having a `bool` and using `&&` which can't overloaded. I'm not sure I understand what you mean by
> with a counter on number of interpolated expressions
I thought maybe using `+`? Technically, `ExpressionContext` is some sort of join-semilattice, but there is no supremum `⋁` operator, so I used `&` for that.

Reply:
So instead of Dynamic or Static, we collect simply collect `num_interpolations`.
If num_interpolations > 0, it means it's dynamic.

Reply:
In a clean design, I think this amounts to passing down a structure through `Reify::into_token_stream` that records this count, so `Reify::into_token_stream(self, &mut Recorder) -> TokenStream`, with `args_used: &mut HashSet<String>` being the prototypical `Recorder` used in `literal`.

[This might also solve a problem of duplicated expressions in nested blocks in the future, if the recorder allows to record expressions and emits them up front.]

An unusual way of a writer-monad for me, coming from functional Haskell, but I'll see if I can make it work :+1:

Reply:
`Reify::into_token_stream(self, &mut Recorder) -> TokenStream`

I agree it's the proper way of doing it.

Although, using a `&Recorder` to collect contextual information might be overkill at the moment, as it only collect a `bool`.

> An unusual way of a writer-monad for me, coming from functional Haskell, but I'll see if I can make it work 👍

I do see a lot of extensive functional language usage (like: iterator) which I was wondering why is it not dealt with using a for-loop. :)